File: src/main/java/com/natu/ftax/transaction/application/TransactionService.java
package com.natu.ftax.transaction.application;

import com.natu.ftax.IDgenerator.domain.IdGenerator;
import com.natu.ftax.common.exception.FunctionalException;
import com.natu.ftax.transaction.domain.Transaction;
import com.natu.ftax.transaction.presentation.EditFieldRequest;
import com.natu.ftax.transaction.presentation.SubmitTransactionRequest;
import org.springframework.stereotype.Service;
import org.springframework.web.multipart.MultipartFile;

import java.util.List;
import java.util.Map;

@Service
public class TransactionService {

    private final IdGenerator idGenerator;
    private final TransactionRepository transactionRepository;
    private final Map<String, PlatformImporter> platformImporters;

    public TransactionService(IdGenerator idGenerator,
            TransactionRepository transactionRepository,
            Map<String, PlatformImporter> platformImporters) {
        this.idGenerator = idGenerator;
        this.transactionRepository = transactionRepository;
        this.platformImporters = platformImporters;

    }

    public Transaction createTransaction() {
        Transaction transaction = Transaction.create(idGenerator.generate());
        transactionRepository.save(transaction);
        return transaction;
    }

    public List<Transaction> getAllTransactions() {
        return transactionRepository.getAllTransactions();
    }

    public void submitDraftTransaction(SubmitTransactionRequest request) {
        Transaction transaction = transactionRepository.getTransactionById(request.getId());
        transaction.submit(request.toCommand());
        transactionRepository.save(transaction);
    }

    public Transaction editField(EditFieldRequest request) {
        Transaction transaction = transactionRepository.getTransactionById(request.getId());
        request.toCommand().execute(transaction);
        transactionRepository.save(transaction);
        return transaction;
    }

    public Transaction editTransaction(String id) {
        Transaction transaction = transactionRepository.getTransactionById(id);
        transaction.edit();
        transactionRepository.save(transaction);
        return transaction;
    }

    public void deleteTransaction(String id) {
        transactionRepository.deleteTransaction(id);
    }

    public void submitDraftTransaction(String id) {
        Transaction transaction = transactionRepository.getTransactionById(id);
        transaction.submit();
        transactionRepository.save(transaction);
    }

    public void importTransactions(String platform, MultipartFile file) {
        PlatformImporter importer = platformImporters.get(
                platform + "Importer");
        if (importer != null) {
            importer.importTransaction(file);
        } else {
            throw new FunctionalException(
                    "Platform " + platform + " not supported");
        }
    }
}

File: src/main/kotlin/com/natu/ftax/transaction/application/TransactionService.kt
package com.natu.ftax.transaction.application

import com.natu.ftax.IDgenerator.domain.IdGenerator
import com.natu.ftax.transaction.domain.Transaction
import com.natu.ftax.transaction.presentation.EditTransactionRequest
import com.natu.ftax.transaction.presentation.SubmitTransactionRequest
import org.springframework.stereotype.Service

@Service
class TransactionService(val idGenerator: IdGenerator, val transactionRepository: TransactionRepository) {

    fun createTransaction(): Transaction {
        val transaction = Transaction.create(idGenerator.generate())
        transactionRepository.save(transaction)
        return transaction
    }

    fun getAllTransactions(): Array<Transaction> {
        return transactionRepository.getAllTransactions()
    }

    fun submitDraftTransaction(request: SubmitTransactionRequest) {
        val transaction = transactionRepository.getTransactionById(request.id)
        transaction.submit(request.toCommand())
        transactionRepository.save(transaction)
    }

    fun editField(request: EditTransactionRequest): Transaction {
        val transaction = transactionRepository.getTransactionById(request.id)
        request.toCommand().execute(transaction)
        transactionRepository.save(transaction)
        return transaction
    }

    fun editTransaction(id: String): Transaction {
        val transaction = transactionRepository.getTransactionById(id)
        transaction.edit()
        transactionRepository.save(transaction)
        return transaction
    }

    fun deleteTransaction(id: String) {
        transactionRepository.deleteTransaction(id)
    }
}